pre-commit

```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+# Usage: copy this file to .git/hooks/
+
+# Exit at first error
+set -Eeu
+
+# For partially committed files, copy the staged changes to a separate location
+TEMPDIR=$(mktemp -d)
+trap "rm -rf $TEMPDIR" EXIT SIGHUP SIGINT SIGQUIT SIGTERM
+git checkout-index --prefix=$TEMPDIR/ -af
+
+GIT_ROOT=$(git rev-parse --show-toplevel)
+
+# Keep using the same target/ directory, not a new one in the temporary
+# directory. This avoids re-parsing everything from scratch every time.
+export CARGO_TARGET_DIR="${GIT_ROOT}/target"
+cd $TEMPDIR
+cargo fmt --check
+if command -v typos >/dev/null; then
+    typos
+fi
+cargo clippy --all -- --deny warnings
```

Comment:
Done!

Reply:
precommits are not strictly github related, but rather just `git` related. `.github` folder deals with projects hosted on github.com and stuff related to that (workflows, issues). IMO it's fine to keep it on top level.

However, I dislike that this would have to be manually copied to `/.git/hooks`, and that's it's a platform specific bash script. I'm a bit opinionated on the topic, and have created a [project](https://github.com/tomkarw/opinionated-rust-template) with what I'd consider best practices. It uses pre-commit python tool and offers a broad range of check and lints out of the box. Here's the relevant bits: [config](https://github.com/tomkarw/opinionated-rust-template/blob/main/template/.pre-commit-config.yaml) and [installation steps](https://github.com/tomkarw/opinionated-rust-template/blob/main/template/justfile#L85-L87).

Could we have a broader discussion about adapting parts of that to buffrs?

Reply:
> precommits are not strictly github related, but rather just git related. .github folder deals with projects hosted on github.com and stuff related to that (workflows, issues).

I agree somewhat, but also this hook is there to mimic github action errors locally. I suggested it as it is our closest "management" folder we have committed in the repository. Im keen on keeping the project a bit condensed / graspable

Im also open for other tools, I think this setup is ok for now though. This is only for the people that want this and is not a strict part of the development environment